File: sdk/src/instructions/deposit.ts
// SPDX-License-Identifier: GPL-3.0
// Copyright (c) 2026 THE3RDWEBLABS (https://github.com/the3rdweblabs)

/**
 * `deposit` - move SPL tokens from the caller's token accounts into the pool
 * vault, crediting `TraderState.deposited_*`.
 *
 * Base and quote move in one instruction; pass `0n` for a side to skip it.
 */
import { PublicKey, TransactionInstruction } from "@solana/web3.js";
import { MAGICLOB_PROGRAM_ID, TOKEN_PROGRAM_ID } from "../constants";
import {
  traderPda,
  vaultBaseTokenPda,
  vaultPda,
  vaultQuoteTokenPda,
} from "../pda";
import type { VaultTransferParams } from "../types";
import {
  DISCRIMINATORS,
  SYSTEM_PROGRAM_ID,
  buildIx,
  ixData,
  mut,
  mutSigner,
  ro,
} from "./shared";

/** Build a `deposit` instruction. */
export function createDepositInstruction(
  params: VaultTransferParams,
  programId: PublicKey = MAGICLOB_PROGRAM_ID
): TransactionInstruction {
  const {
    market,
    authority,
    baseMint,
    quoteMint,
    baseTokenAccount,
    quoteTokenAccount,
    baseAmount = 0n,
    quoteAmount = 0n,
  } = params;

  const data = ixData(DISCRIMINATORS.deposit)
    .u64(baseAmount)
    .u64(quoteAmount)
    .toBuffer();

  return buildIx(
    [
      mutSigner(authority),
      mut(market),
      mut(vaultPda(market, programId).address),
      mut(traderPda(market, authority, programId).address),
      ro(baseMint),
      ro(quoteMint),
      mut(baseTokenAccount),
      mut(quoteTokenAccount),
      mut(vaultBaseTokenPda(market, programId).address),
      mut(vaultQuoteTokenPda(market, programId).address),
      ro(TOKEN_PROGRAM_ID),
      ro(SYSTEM_PROGRAM_ID),
    ],
    data,
    programId
  );
}

File: contracts/programs/magiclob/src/instructions/deposit.rs
// SPDX-License-Identifier: GPL-3.0
// Copyright (c) 2026 THE3RDWEBLABS (https://github.com/the3rdweblabs)

//! Deposit SPL tokens into the pool vault.

use crate::errors::MagiCLOBError;
use crate::state::*;
use anchor_lang::prelude::*;
use anchor_spl::token::{self, Token, TokenAccount, Transfer};

#[derive(Accounts)]
pub struct Deposit<'info> {
    #[account(mut)]
    pub authority: Signer<'info>,
    #[account(mut)]
    pub market: Box<Account<'info, MarketState>>,
    #[account(
        mut,
        seeds = [VAULT_SEED, market.key().as_ref()],
        bump = vault.bump,
        constraint = vault.market == market.key() @ MagiCLOBError::InvalidAuthority
    )]
    pub vault: Box<Account<'info, VaultState>>,
    #[account(
        mut,
        seeds = [TRADER_SEED, market.key().as_ref(), authority.key().as_ref()],
        bump = trader.bump,
        constraint = trader.owner == authority.key() @ MagiCLOBError::TraderAuthMismatch
    )]
    pub trader: Box<Account<'info, TraderState>>,
    pub base_mint: Box<Account<'info, anchor_spl::token::Mint>>,
    pub quote_mint: Box<Account<'info, anchor_spl::token::Mint>>,
    #[account(
        mut,
        constraint = base_token_account.mint == base_mint.key() @ MagiCLOBError::MintMismatch,
        constraint = base_token_account.owner == authority.key() @ MagiCLOBError::InvalidTokenAccountOwner
    )]
    pub base_token_account: Box<Account<'info, TokenAccount>>,
    #[account(
        mut,
        constraint = quote_token_account.mint == quote_mint.key() @ MagiCLOBError::MintMismatch,
        constraint = quote_token_account.owner == authority.key() @ MagiCLOBError::InvalidTokenAccountOwner
    )]
    pub quote_token_account: Box<Account<'info, TokenAccount>>,
    #[account(
        mut,
        seeds = [VAULT_SEED, market.key().as_ref(), b"base"],
        bump,
        constraint = base_vault_account.owner == vault.key() @ MagiCLOBError::InvalidTokenAccountOwner,
        constraint = base_vault_account.mint == base_mint.key() @ MagiCLOBError::MintMismatch
    )]
    pub base_vault_account: Box<Account<'info, TokenAccount>>,
    #[account(
        mut,
        seeds = [VAULT_SEED, market.key().as_ref(), b"quote"],
        bump,
        constraint = quote_vault_account.owner == vault.key() @ MagiCLOBError::InvalidTokenAccountOwner,
        constraint = quote_vault_account.mint == quote_mint.key() @ MagiCLOBError::MintMismatch
    )]
    pub quote_vault_account: Box<Account<'info, TokenAccount>>,
    pub token_program: Program<'info, Token>,
    pub system_program: Program<'info, System>,
}

pub fn deposit(ctx: Context<Deposit>, base_amount: u64, quote_amount: u64) -> Result<()> {
    let market = &ctx.accounts.market;
    crate::instructions::ensure_active(market)?;

    if base_amount > 0 {
        token::transfer(
            CpiContext::new(
                ctx.accounts.token_program.key(),
                Transfer {
                    from: ctx.accounts.base_token_account.to_account_info(),
                    to: ctx.accounts.base_vault_account.to_account_info(),
                    authority: ctx.accounts.authority.to_account_info(),
                },
            ),
            base_amount,
        )?;
        ctx.accounts.trader.deposited_base = ctx
            .accounts
            .trader
            .deposited_base
            .checked_add(base_amount)
            .ok_or(MagiCLOBError::ArithmeticOverflow)?;
        ctx.accounts.trader.base_balance = ctx
            .accounts
            .trader
            .base_balance
            .checked_add(base_amount)
            .ok_or(MagiCLOBError::ArithmeticOverflow)?;
    }

    if quote_amount > 0 {
        token::transfer(
            CpiContext::new(
                ctx.accounts.token_program.key(),
                Transfer {
                    from: ctx.accounts.quote_token_account.to_account_info(),
                    to: ctx.accounts.quote_vault_account.to_account_info(),
                    authority: ctx.accounts.authority.to_account_info(),
                },
            ),
            quote_amount,
        )?;
        ctx.accounts.trader.deposited_quote = ctx
            .accounts
            .trader
            .deposited_quote
            .checked_add(quote_amount)
            .ok_or(MagiCLOBError::ArithmeticOverflow)?;
        ctx.accounts.trader.quote_balance = ctx
            .accounts
            .trader
            .quote_balance
            .checked_add(quote_amount)
            .ok_or(MagiCLOBError::ArithmeticOverflow)?;
    }

    emit!(crate::instructions::DepositEvent {
        market: market.key(),
        owner: ctx.accounts.authority.key(),
        base_amount,
        quote_amount,
    });

    Ok(())
}
